### ISSUE TITLE
systemd-networkd fails to setup host0 container interface

### ISSUE THREAD

**Issue by tchernomax** [systemd/systemd#6733]
### Submission type

 - Bug report

### systemd version the issue has been seen with

systemd 234
+PAM -AUDIT -SELINUX -IMA -APPARMOR +SMACK -SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD -IDN2 +IDN default-hierarchy=hybrid

### Used distribution

Archlinux

### Expected behaviour you didn't see

srv-01 is Archlinux based (`sudo pacstrap -c -i -d /var/lib/machines/srv-init base`) systemd-nspawn container started with `sudo systemd-nspawn -M srv-01 -x -n -b -D /var/lib/machines/srv-init`

/etc/systemd/network/lan.network config is applied.

```console
[root@srv-01 ~]# ip link show host0
2: host0@if12: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP mode DEFAULT group default qlen 1000
    link/ether ee:9a:53:f0:13:3b brd ff:ff:ff:ff:ff:ff link-netnsid 0

[root@srv-01 ~]# cat /etc/systemd/network/lan.network
[Match]
MACAddress=ee:9a:53:f0:13:3b
[Network]
DHCP=no
Address=192.168.2.2/24
Gateway=192.168.2.1

[root@srv-01 ~]# systemctl is-enabled systemd-networkd
enabled

[root@srv-01 ~]# reboot
…

[root@srv-01 ~]# ip -4 addr show host0
2: host0@if12: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    inet 192.168.2.2/24 brd 192.168.2.255 scope link host0
       valid_lft forever preferred_lft forever

```

### Unexpected behaviour you saw

srv-01 is Archlinux based (`sudo pacstrap -c -i -d /var/lib/machines/srv-init base`) systemd-nspawn container started with `sudo systemd-nspawn -M srv-01 -x -n -b -D /var/lib/machines/srv-init`

systemd-networkd doesn't apply the /etc/systemd/network/lan.network config.
host0 end up with a 169.254.0.0/16 ip.

```console
[root@srv-01 ~]# ip link show host0
2: host0@if12: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP mode DEFAULT group default qlen 1000
    link/ether ee:9a:53:f0:13:3b brd ff:ff:ff:ff:ff:ff link-netnsid 0

[root@srv-01 ~]# cat /etc/systemd/network/lan.network
[Match]
MACAddress=ee:9a:53:f0:13:3b
[Network]
DHCP=no
Address=192.168.2.2/24
Gateway=192.168.2.1

[root@srv-01 ~]# systemctl is-enabled systemd-networkd
enabled

[root@srv-01 ~]# reboot
…

[root@srv-01 ~]# ip -4 addr show host0
2: host0@if12: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    inet 169.254.215.95/16 brd 169.254.255.255 scope link host0
       valid_lft forever preferred_lft forever

[root@srv-01 ~]# networkctl
IDX LINK             TYPE               OPERATIONAL SETUP
  1 lo               loopback           routable    configured
  2 host0            ether              degraded    configuring

2 links listed.

[root@srv-01 ~]# journalctl -b 1 -u systemd-networkd
-- Logs begin at Sun 2017-09-03 12:54:05 CEST, end at Sun 2017-09-03 13:15:30 CEST. --
Sep 03 12:54:31 srv-01 systemd[1]: Starting Network Service...
Sep 03 12:54:31 srv-01 systemd-networkd[123]: host0: Gained IPv6LL
Sep 03 12:54:31 srv-01 systemd-networkd[123]: Enumeration completed
Sep 03 12:54:31 srv-01 systemd[1]: Started Network Service.
Sep 03 12:54:31 srv-01 systemd-networkd[123]: lo: Configured
Sep 03 12:58:22 srv-01 systemd[1]: Stopping Network Service...
Sep 03 12:58:22 srv-01 systemd[1]: Stopped Network Service.
Sep 03 12:58:22 srv-01 systemd[1]: Starting Network Service...
Sep 03 12:58:22 srv-01 systemd-networkd[152]: host0: Gained IPv6LL
Sep 03 12:58:22 srv-01 systemd-networkd[152]: Enumeration completed
Sep 03 12:58:22 srv-01 systemd[1]: Started Network Service.
Sep 03 12:58:22 srv-01 systemd-networkd[152]: lo: Configured
Sep 03 12:59:03 srv-01 systemd[1]: Stopping Network Service...
Sep 03 12:59:03 srv-01 systemd[1]: Stopped Network Service.

[root@srv-01 ~]# systemctl restart systemd-networkd
<same result>
```

### Steps to reproduce the problem

On Archlinux:
```console
user@host % sudo btrfs subvolume create /var/lib/machines/test
…
user@host % sudo pacstrap -c -i -d /var/lib/machines/test base
…
user@host % sudo systemd-nspawn -M container -x -n -b -D /var/lib/machines/test
…
container login: root
[root@container ~]# cat > /etc/systemd/network/host0.network
[Match]
Name=host0
[Network]
DHCP=no
Address=192.168.2.2/24

user@host % sudo ip link set up dev ve-container

[root@container ~]# ip -4 addr show dev host0
[root@container ~]# systemctl start systemd-networkd
[root@container ~]# <wait>
[root@container ~]# ip -4 addr show dev host0
2: host0@if14: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000 link-netnsid 0
    inet 169.254.142.9/16 brd 169.254.255.255 scope link host0
       valid_lft forever preferred_lft forever
[root@container ~]# networkctl status host0
* 2: host0
       Link File: n/a
    Network File: /usr/lib/systemd/network/80-container-host0.network
            Type: ether
           State: degraded (configuring)
      HW Address: 7a:c9:b2:a0:e0:5f
         Address: 169.254.142.9
                  fe80::78c9:b2ff:fea0:e05f
```

**Comment by akarl10** [comment]
I had a similar issue with ubuntu artful.
as you can see /usr/lib/systemd/network/80-container-host0.network is applied
if you read man systemd.network you see the reason:

`       The .network files are read from the files located in the system network directory /lib/systemd/network, the volatile runtime network directory /run/systemd/network and the local administration
       network directory /etc/systemd/network. All configuration files are collectively sorted and processed in lexical order, regardless of the directories in which they live. However, files with
       identical filenames replace each other. Files in /etc have the highest priority, files in /run take precedence over files with the same name in /lib. This can be used to override a
       system-supplied configuration file with a local file if needed. As a special case, an empty file (file size 0) or symlink with the same name pointing to /dev/null disables the configuration file
       entirely (it is "masked").`

In short: naming the .network file 00-lan.network (or 80-container-host0.network to explicitly override the default configuration) should resolve your problem

**Comment by tchernomax** [comment]
@akarl10 it works ; thanks A LOT for that quick answer/solution !

So it's not a bug (sorry for the noise) ; I close this issue.
